receiver/vcenterreceiver/scraper.go

```diff
@@ -0,0 +1,305 @@
+	)
+}
+
+func (v *vcenterMetricScraper) collectHosts(
+	ctx context.Context,
+	colTime pcommon.Timestamp,
+	cluster *object.ClusterComputeResource,
+	errs *scrapererror.ScrapeErrors,
+) {
+	hosts, err := cluster.Hosts(ctx)
+	if err != nil {
+		errs.AddPartial(1, err)
+		return
+	}
+
+	for _, h := range hosts {
+		v.collectHost(ctx, colTime, h, cluster, errs)
+	}
+}
+
+func (v *vcenterMetricScraper) collectHost(
+	ctx context.Context,
+	now pcommon.Timestamp,
+	host *object.HostSystem,
+	cluster *object.ClusterComputeResource,
+	errs *scrapererror.ScrapeErrors,
+) {
+	var hwSum mo.HostSystem
+	err := host.Properties(ctx, host.Reference(),
+		[]string{
+			"config",
+			"summary.hardware",
+			"summary.quickStats",
+		}, &hwSum)
+
+	if err != nil {
+		errs.AddPartial(1, err)
+		return
+	}
+	v.recordHostSystemMemoryUsage(now, hwSum)
+	v.recordHostPerformanceMetrics(ctx, hwSum, errs)
+	v.mb.EmitForResource(
+		metadata.WithVcenterHostName(host.Name()),
+		metadata.WithVcenterClusterName(cluster.Name()),
+	)
+}
+
+func (v *vcenterMetricScraper) collectResourcePools(
+	ctx context.Context,
+	ts pcommon.Timestamp,
+	errs *scrapererror.ScrapeErrors,
+) {
+	rps, err := v.client.ResourcePools(ctx)
+	if err != nil {
+		errs.AddPartial(1, err)
+		return
+	}
+	for _, rp := range rps {
+		var moRP mo.ResourcePool
+		err = rp.Properties(ctx, rp.Reference(), []string{
+			"summary",
+			"summary.quickStats",
+			"name",
+		}, &moRP)
+		if err != nil {
+			errs.AddPartial(1, err)
+			continue
+		}
+		v.recordResourcePool(ts, moRP)
+		v.mb.EmitForResource(metadata.WithVcenterResourcePoolName(rp.Name()))
+	}
+}
+
+func (v *vcenterMetricScraper) collectVMs(
+	ctx context.Context,
+	colTime pcommon.Timestamp,
+	cluster *object.ClusterComputeResource,
+	errs *scrapererror.ScrapeErrors,
+) {
+	vms, err := v.client.VMs(ctx)
+	if err != nil {
+		errs.AddPartial(1, err)
+		return
+	}
+	poweredOffVMs := 0
+	for _, vm := range vms {
+		var moVM mo.VirtualMachine
+		err := vm.Properties(ctx, vm.Reference(), []string{
+			"config",
+			"runtime",
+			"summary",
+		}, &moVM)
+
+		if err != nil {
+			errs.AddPartial(1, err)
+			continue
+		}
+
+		vmUUID := moVM.Config.InstanceUuid
+
+		if string(moVM.Runtime.PowerState) == "poweredOff" {
+			poweredOffVMs++
+		}
+
+		host, err := vm.HostSystem(ctx)
+		if err != nil {
+			errs.AddPartial(1, err)
+			return
+		}
+		hostname, err := host.ObjectName(ctx)
+		if err != nil {
+			errs.AddPartial(1, err)
+			return
+		}
+
+		v.collectVM(ctx, colTime, moVM, errs)
+		v.mb.EmitForResource(
+			metadata.WithVcenterVMName(vm.Name()),
+			metadata.WithVcenterVMID(vmUUID),
+			metadata.WithVcenterClusterName(cluster.Name()),
+			metadata.WithVcenterHostName(hostname),
+		)
+	}
+
+	v.mb.RecordVcenterClusterVMCountDataPoint(colTime, int64(len(vms))-int64(poweredOffVMs), metadata.AttributeVMCountPowerStateOn)
+	v.mb.RecordVcenterClusterVMCountDataPoint(colTime, int64(poweredOffVMs), metadata.AttributeVMCountPowerStateOff)
```

Comment:
What do you think about returning these counts from this function, and passing them into `collectCluster`? The way this is currently done is a bit too tricky because the next call to `EmitForResource` is in another function. It would be very easy to change up the order of functions without realizing that these metrics are then attached to the wrong resource.